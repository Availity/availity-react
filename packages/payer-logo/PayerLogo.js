import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useEffectAsync } from '@availity/hooks';
import { avSlotMachineApi } from '@availity/api-axios';
import get from 'lodash.get';

const spaceIDQuery = `
query($id: ID!){
  space(id: $id){
    images{
      name
      value
    }
  }
}
`;

const payerIDQuery = `
query($payerIDs: [String!], $types: [String!]){
  spaces(payerIDs: $payerIDs, types: $types){
    spaces{
      images{
        name
        value
      }
    }
  }
}
`;

const fetchLogo = async (query, variables, path, clientId) => {
  try {
    const {
      data: { data },
    } = await avSlotMachineApi.create(
      {
        query,
        variables,
      },
      { headers: { 'X-Client-ID': clientId } }
    );

    const images = get(data, path, []).reduce((accum, { name, value }) => {
      accum[name] = value;
      return accum;
    }, {});

    return images.logo;
  } catch (error) {
    return error;
  }
};

export const getLogo = async (spaceId, payerId, clientId) => {
  if (!clientId) {
    throw new Error('clientId is required');
  }

  try {
    let url;
    if (spaceId) {
      const variables = { id: spaceId };
      const path = 'space.images';
      url = await fetchLogo(spaceIDQuery, variables, path, clientId);
    } else if (payerId) {
      const variables = { payerIDs: [payerId], types: ['space'] };
      const path = 'spaces.spaces[0].images';
      url = await fetchLogo(payerIDQuery, variables, path, clientId);

      // We can probably remove this at some point once our spaces data is complete
      if (!url) {
        url = `/public/apps/eligibility/images/value-add-logos/${payerId.replace(
          /\s/g,
          ''
        )}.gif`;
      }
    }

    return url;
  } catch (error) {
    return error;
  }
};

const PayerLogo = ({ spaceId, payerId, clientId, ...props }) => {
  const [url, setUrl] = useState(null);

  useEffectAsync(async () => {
    const _url = await getLogo(spaceId, payerId, clientId);
    setUrl(_url);
  }, [spaceId, payerId, clientId]);

  if (!clientId || (!payerId && !spaceId)) return null;

  return <img src={url} alt="Payer logo" {...props} />;
};

PayerLogo.propTypes = {
  spaceId: PropTypes.string,
  payerId: PropTypes.string,
  clientId: PropTypes.string.isRequired,
};

export default PayerLogo;
