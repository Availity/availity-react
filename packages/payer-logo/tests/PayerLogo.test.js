import React from 'react';
import { render, waitForElement, cleanup } from '@testing-library/react';
import { avSlotMachineApi } from '@availity/api-axios';
import PayerLogo, { getLogo } from '../PayerLogo';

jest.mock('@availity/api-axios');

afterEach(cleanup);

describe('PayerLogo', () => {
  test('should not render when no space or payer id', () => {
    const { container } = render(<PayerLogo clientId="my-client-id" />);

    expect(container).toMatchSnapshot();
  });

  test('should render with payer id', async () => {
    avSlotMachineApi.create.mockResolvedValue({
      data: {
        data: {
          spaces: {
            spaces: [
              {
                images: [
                  {
                    name: 'logo',
                    value:
                      '/static/spaces/73162546201440710195134200002269/banner.png',
                  },
                ],
              },
            ],
          },
        },
      },
    });

    const { container } = render(
      <PayerLogo payerId="BCBSF" clientId="my-client-id" />
    );

    await waitForElement(() =>
      container.querySelector(
        'img[src="/static/spaces/73162546201440710195134200002269/banner.png"]'
      )
    );

    expect(container).toMatchSnapshot();
  });

  test('should render with default when payer id returns no results', async () => {
    avSlotMachineApi.create.mockResolvedValue({
      data: {
        data: {
          spaces: {
            spaces: [],
          },
        },
      },
    });

    const { container } = render(
      <PayerLogo payerId="00681" clientId="my-client-id" />
    );

    await waitForElement(() =>
      container.querySelector(
        'img[src="/public/apps/eligibility/images/value-add-logos/00681.gif"]'
      )
    );

    expect(container).toMatchSnapshot();
  });

  test('should render with space id', async () => {
    avSlotMachineApi.create.mockResolvedValue({
      data: {
        data: {
          space: {
            images: [
              {
                name: 'logo',
                value:
                  '/static/spaces/73162546201441126239486200007187/banner.png',
              },
            ],
          },
        },
      },
    });

    const { container } = render(
      <PayerLogo
        spaceId="73162546201441126239486200007187"
        clientId="my-client-id"
      />
    );

    await waitForElement(() =>
      container.querySelector(
        'img[src="/static/spaces/73162546201441126239486200007187/banner.png"]'
      )
    );

    expect(container).toMatchSnapshot();
  });

  test('should throw error on missing clientId', async () => {
    let message;
    try {
      await getLogo();
    } catch (error) {
      const { message: mess } = error;
      message = mess;
    }

    expect(message).toBe('clientId is required');
  });

  test('should return error on rejected promise', async () => {
    avSlotMachineApi.create.mockRejectedValue('This field was rejected');

    const response = await getLogo(null, '3', 'test-client-id');

    expect(response).toBe('This field was rejected');
  });
});
