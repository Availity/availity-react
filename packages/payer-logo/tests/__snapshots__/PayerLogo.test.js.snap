// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayerLogo should not render when no space or payer id 1`] = `<div />`;

exports[`PayerLogo should render with payer id 1`] = `
<div>
  <img
    alt="Payer logo"
    src="/static/spaces/73162546201440710195134200002269/banner.png"
  />
</div>
`;

exports[`PayerLogo should render with default when payer id returns no results 1`] = `
<div>
  <img
    alt="Payer logo"
    src="/public/apps/eligibility/images/value-add-logos/00681.gif"
  />
</div>
`;

exports[`PayerLogo should render with space id 1`] = `
<div>
  <img
    alt="Payer logo"
    src="/static/spaces/73162546201441126239486200007187/banner.png"
  />
</div>
`;
